File: src/main/java/dev/elysium/eracescast/client/ERacesPayload.java
package dev.elysium.eracescast.client;

import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.network.codec.PacketCodecs;
import net.minecraft.network.packet.CustomPayload;
import net.minecraft.util.Identifier;

public record ERacesPayload(Long timestamp, String type, String payload) implements CustomPayload {
    public static final CustomPayload.Id<ERacesPayload> ID = new CustomPayload.Id<>(Identifier.of("elysium", "eraces_cast"));

    public static final PacketCodec<RegistryByteBuf, ERacesPayload> CODEC = PacketCodec.tuple(
            PacketCodecs.LONG, ERacesPayload::timestamp,
            PacketCodecs.STRING, ERacesPayload::type,
            PacketCodecs.STRING, ERacesPayload::payload,
            ERacesPayload::new
    );

    @Override
    public CustomPayload.Id<? extends CustomPayload> getId() {
        return ID;
    }
}

File: src/main/kotlin/dev/elysium/eracescast/packets/ERacesPayload.kt
package dev.elysium.eracescast.packets

import dev.elysium.eracescast.utils.UTFEncoder
import net.minecraft.network.RegistryByteBuf
import net.minecraft.network.codec.PacketCodec
import net.minecraft.network.packet.CustomPayload
import net.minecraft.util.Identifier

data class ERacesPayload(
    val type: String,
    val payload: String
) : CustomPayload {

    companion object {
        val ID: CustomPayload.Id<ERacesPayload> =
            CustomPayload.Id(Identifier.of("elysium", "eraces_cast"))

        val CODEC: PacketCodec<RegistryByteBuf, ERacesPayload> = PacketCodec.tuple(
            UTFEncoder, ERacesPayload::type,
            UTFEncoder, ERacesPayload::payload,
            ::ERacesPayload
        )
    }

    override fun getId(): CustomPayload.Id<out CustomPayload> = ID
}
